proposals/filecoin-pinning.md

```diff
@@ -0,0 +1,107 @@
+# JS-driven storing from IPFS-API to Filecoin
+
+Authors: @willscott
+
+Initial PR: https://github.com/protocol/web3-dev-team/pull/69
+
+
+## Purpose &amp; impact 
+#### Background &amp; intent
+<!--
+Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives. 
+-->
+
+Currently, to make sure data is persistently stored in a decentralized mechanism as an IPFS user is a complicated process.
+In this position, you need to get access to a filecoin node, either by running a full lotus node, or paying someone else, like
+infura to run one for you. Then, on the lotus command line, you can make a 'deal' with a miner to store data.
+The lotus deal process allows the source of the data to be a local IPFS node, and will then retreave and stream the data
+to the miner for storage, but there's no way to 'push' this data to the filecoin system from IPFS even though doing
+so is conceptually similar to the existing 'remote pinning' API that does already exist for centralized pinning providers.
+
+#### Assumptions &amp; hypotheses
+
+* IPFS users wish to store data persistently in filecoin
+
+#### User workflow example
+
+Alice has been using [Pathephone](https://pathephone.github.io/) as a desktop music player with an IPFS database (or a similar dapp) and
+wants to store the data longer term. Alice searches for how to 'backup IPFS music' and finds a web page that walks her through the process.
+
+This page asks her to link her filecoin wallet, or allows her to transfer filecoin to a local wallet address generated within the browser session.
+It then makes a deal with a miner and backs data she selects up to filecoin.
+
+#### Impact
+High. There's a gap with no 'golden path' between ipfs and filecoin at the moment
+
+<!--
+Explain why you have chosen this rating
+What awesome potential impact/outcomes/results will we see if we nail this project?
+-->
+
+#### Leverage
+Medium. The core logic can serve as a re-usable block for other adaptions and scenarios - e.g. ones wehre a developer pays on behalf of a user, or a more full-featured dapp that makes use of backup within itself.
+
+<!-- Explain the opportunity or leverage point for our subsequent velocity/impact (e.g. by speeding up development, enabling more contributors, etc)
+-->
+
+#### Confidence
+_How sure are we that this impact would be realized? Label from [this scale](https://medium.com/@nimay/inside-product-introduction-to-feature-priority-using-ice-impact-confidence-ease-and-gist-5180434e5b15)_.
+
+<!--Explain why this rating-->
+1-3 - we know there should be something. we haven't heard this approach directly asked for.
+
+
+## Project definition
+#### Brief plan of attack
+
+A core JS library that uses the local IPFS library to
```

Comment:
What would the changes to this flow look like if we were trying to support enabling someone to go from a pinning service to Filecoin? 

A common pattern is developers will pin content in Pinata on behalf of users (common in the NFT space), and so as a user you may not have a local IPFS node running

Reply:
I think this depends on the specific flow.
If this is a user with a wallet who is initiating the pin of content already in a pinning service, then a similar design as proposed here could work. Instead of referencing the peer id of the local ipfs node for the miner to pull content from, the content would be available from the pinning service, and the miner could pull it from there, so it essentially would be a couple steps the user's JS library doesn't need to do in that case.

If it is the developer who has previously had the account with the pinning service, and is now wanting to pay for the filecoin storage on behalf of their user as the primary use case we're supporting, we would likely want a different setup. 

Do you have insight into what the more important setup in terms of where the money is coming from we should be thinking about?

Reply:
I think the ideal flow is as you described- so I think the base pattern you're describing makes sense. The main barrier to this being the flow will end up being the cost and whatever other pathways are encouraged in the short term (e.g. batching solutions for small files) -> though I think we should be thinking of that as a separate flow